### PR TITLE
Prevent chunk griefing by sampling random chunk-parts

### PR DESCRIPTION
**Pull request by Akashnil**
Every ShardsManager requests some random chunk parts in addition to those owned by it, from their owners. If a chunk is to be included in a block, these additional chunks must have been been received. A local_random_seed is used to sample them.

TODO:
Test that chuck griefing is not possible with these changes.

### CODE REVIEW

File: chain/chunks/src/lib.rs [hunk 1]
```diff
@@ -248,7 +258,9 @@ impl ShardsManager {
                 true
             } else {
                 if let Some(me) = &self.me {
-                    &self.runtime_adapter.get_part_owner(&parent_hash, part_ord)? == me
+                    // Request all parts that are owned by me and some small number of sampled parts.
+                    (&self.runtime_adapter.get_part_owner(&parent_hash, part_ord)? == me) |
```
Comment (SkidanovAlex) [inline chain/chunks/src/lib.rs]: `|`  -> `||` (there are several places like that)

File: chain/chunks/src/lib.rs [hunk 2]
```diff
@@ -870,6 +898,7 @@ impl ShardsManager {
                 .parts
                 .iter()
                 .filter_map(|(part_ord, part_entry)| {
+                    // Should we include sampled parts here too??
```
Comment (SkidanovAlex) [inline chain/chunks/src/lib.rs]: No, here we only keep parts that someone can request from us in the future, knowing that we own them. So no need to save sampled parts here.

File: chain/chunks/src/lib.rs [hunk 3]
```diff
@@ -780,6 +792,21 @@ impl ShardsManager {
...
+    fn should_sample_part(&self, prev_block_hash: &CryptoHash, part_ord: u64) -> Result<bool, Error> {
+        let mut s = DefaultHasher::new();
+        let num_total_parts = self.runtime_adapter.num_total_parts(&prev_block_hash) as u64;
+        self.local_random_seed.hash(&mut s);
+        prev_block_hash.hash(&mut s);
+        for _ in 1..NUM_PARTS_TO_BE_SAMPLED {
+            if part_ord == s.finish() % num_total_parts {
+                return Ok(true)
+            }
+            s.write_i8(0); // Just advance the prng's state forward to get another value.
```
Comment (SkidanovAlex) [inline chain/chunks/src/lib.rs]: you can get the same `part_ord` for both parts we sample, but I think that's fine.

Comment (Akashnil) [reply chain/chunks/src/lib.rs]: True. It can also become a part that is owned by me. Relying on randomness to deter attackers enough of the time.

**Comment by nearmax** [comment]
CI is failing probably because you have not updated `near/res/testnet.json`

**Pull request closed without merging**